File: webots/controllers/python/pids/position.py
from . pid import PID


class PositionController:
    '''
     Demand is input as desired speed in meter per second, output as
     angles in degrees.
    '''

    KP = 25

    def __init__(self):

        self.roll_pid = PID()

        self.pitch_pid = PID()

    def run(self, dx, dy, dt, demands):

        thrust, roll, pitch, yaw = demands

        roll = self.run_axis(self.roll_pid, roll, dt, dy)

        pitch = self.run_axis(self.pitch_pid, pitch, dt, dx)

        return thrust, roll, pitch, yaw

    def run_axis(self, pid, demand, dt, actual):

        return pid.run_p(self.KP, dt, demand, actual)

File: haskell/pids/Position.hs
{--
  X/Y position PID control algorithm for real and simulated flight controllers
 
  Copyright (C) 2024 Simon D. Levy
 
  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, in version 3.
 
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.
 
  You should have received a copy of the GNU General Public License
  along with this program. If not, see <http://www.gnu.org/licenses/>.
--} 

{-# LANGUAGE DataKinds #-}
{-# LANGUAGE RebindableSyntax #-}

module Position where

import Language.Copilot
import Copilot.Compile.C99

import Demands
import State
import Utils

run kp dt target actual = demand where

  error = target - actual

  demand = kp * error

{--
  Demand is input as desired speed in meter per second, output as
  angles in degrees.
--}

positionController dt state demands = demands'  where

  kp = 25
    
  rollDemand = run kp dt (roll demands) (dy state)

  pitchDemand = run kp dt (pitch demands) (dx state)

  demands' = Demands (thrust demands) rollDemand pitchDemand (yaw demands)
